### PR TITLE
feat(desktop): cloud sessions — settings opt-in, GitHub onboarding flow, and UX polish

### PR DESCRIPTION
**Pull request by saoudrizwan**
### Related Issue

Builds on JC's first pass at desktop cloud sessions ([`codex/desktop-cloud-agents`](https://github.com/cline/cline/compare/main...codex/desktop-cloud-agents)). This branch includes all of that work, merged with the latest `main`, plus productization changes to make the feature shippable.

### Description

This PR takes the cloud sessions first pass and gets it ready for real users:

**1. Explicit opt-in toggle in Settings (replaces the PostHog rollout flag)**

Cloud sessions are in preview, so instead of gating on the remote `code-cloud-agents` flag, users now opt in via a "Cloud sessions" (Preview) toggle in Settings > General. The setting persists in a desktop-owned `settings/code-settings.json` file, deliberately kept out of the shared `global-settings.json` (whose strict schema means an older CLI writer would silently strip unknown fields). Toggling broadcasts a `feature_flags_changed` event so every open composer reacts immediately, no restart needed; turning it off snaps an unsent cloud composer back to Local. The `CLINE_CODE_CLOUD_AGENTS` env override still wins in both directions for development.

**2. First-run onboarding panel for the GitHub connect flow**

Previously, a new user who clicked "Cloud" without GitHub connected saw two empty pickers, a disabled composer, and the only hint buried inside the repo-picker dropdown. Now the composer is replaced by an onboarding panel (`cloud-onboarding.tsx`) that:

- explains what cloud sessions are (isolated sandbox, works on a branch, keeps running after you close the app)
- walks through three visual steps: Connect GitHub, Pick your repositories, Start a session
- opens the correct dashboard integrations URL (personal vs org, from the API's `connectUrl`)
- auto-detects completion: the panel polls the repository list every 6s and refetches on window focus, so when the user finishes the dashboard hand-off in their browser and comes back, it flips into the ready composer on its own
- has dedicated variants for signed-out (sign-in CTA plus feature pitch), GitHub connected but zero repositories accessible (explains granting the GitHub App repo access, with a "Manage repository access" button), and network errors (retry)

**3. Smaller UX fixes**

- Sessions view context menu now offers Rename for cloud sessions (the sidecar and sidebar/chat header already supported it; the menu was the odd one out)
- A one-line teaching hint under the ready cloud composer: "Cloud sessions run on a secure sandbox, work on a branch, and keep going even when you close the app."
- Merge-conflict resolution unified `main`'s auth-refresh telemetry with this branch's extracted `cline-auth.ts` (the `workos:` token prefix now applies on the account path too, with the test updated to match)

Deliberately out of scope (noted for follow-ups): billing-specific error UX (needs a typed error code from the API), cancel-during-provisioning, in-app GitHub App install, attachments/fork/checkpoints for cloud sessions, and docs (no Cloud Sessions docs page exists yet to deep-link).

### Test Procedure

- `bun run typecheck` in `apps/examples/desktop-app` passes.
- Full desktop vitest suite: 59/61 files pass, 596+ tests. The two failures (`scripts/generate-update-manifest.test.ts`, `webview/styles/theme-integration.test.ts`) reproduce identically on plain `origin/main` in this environment and are unrelated.
- New unit tests: `sidecar/desktop-settings.test.ts` (persistence, defaults, malformed files) and updated `sidecar/feature-flags.test.ts` (toggle gating, env override precedence).
- Manual end-to-end test against the real Cline API using Saoud's account (which has no GitHub connected, i.e. the exact new-user path): verified cloud UI hidden by default, toggle appears in Settings, enabling it reveals Local/Cloud live without a restart, clicking Cloud shows the onboarding panel, and disabling the toggle restores the local composer.
- The connect-completion journey (onboarding panel auto-flipping to the ready composer, repo/branch pickers populating, send enabling) was exercised through a local proxy to the Cline API that simulates the GitHub integration becoming connected, since actually connecting GitHub requires a browser OAuth hand-off.

### Type of Change

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [x] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`bun test`) and code is formatted and linted (`bun run format && bun run lint`)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

Settings > General with the new Cloud sessions toggle:

Cloud sessions toggle in Settings

Onboarding panel shown when Cloud is selected without GitHub connected:

Cloud onboarding panel

Ready cloud composer after connecting, with repo and branch selected and the teaching hint:

Ready cloud composer

### Additional Notes

The UX review scratchpad with the full ranked gap list from dogfooding JC's branch is attached to the agent run as `cloud_sessions_ux_scratchpad.md`.

### CODE REVIEW

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 1]
```diff
@@ -0,0 +1,2154 @@
...
+			Boolean(streamed) &&
+			snapshotAssistantTexts.some((text) => text.includes(streamed));
+		for (const event of segment) {
+			if (contentPersisted && SUPERSEDABLE_CONTENT_EVENTS.has(event.event)) {
+				continue;
+			}
+			// The separately fetched queue snapshot is newer than buffered copies.
+			if (event.event === "session.pending_prompts") {
+				continue;
+			}
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts]: Fixed in b091c72e4. reconcileBufferedCloudEvents now takes a queueSnapshotApplied option: when rehydration's session.pending_prompts fetch fails, the newest buffered queue snapshot is replayed instead of discarded, so queued and steered prompts survive the reconnect. Added a unit test covering the failed-fetch path.

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 2]
```diff
@@ -0,0 +1,2154 @@
...
+							// The backend may resolve an omitted branch to the repo default,
+							// so only require a match when we asked for a specific one.
+							(!requestedBranch ||
+								session.repoContext.branch === requestedBranch) &&
+							Date.parse(session.createdAt) >= requestedAt - 5_000,
+					)
+					.sort(
+						(left, right) =>
+							Date.parse(right.createdAt) - Date.parse(left.createdAt),
+					)[0];
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts]: Fixed in b091c72e4. CloudSessionApi now records session ids claimed by create() calls in this process, for both successful creates and timeout recoveries. Recovery skips already-claimed records, so overlapping identical requests adopt distinct sessions, and a request whose only candidate is already owned surfaces its original error instead of attaching to another request's session. Added two unit tests covering both scenarios.

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 3]
```diff
@@ -0,0 +1,2154 @@
...
+
+	async create(input: CreateCloudSessionInput): Promise<JsonRecord> {
+		if (this.disposed) {
+			throw new Error("Cloud session manager was disposed");
+		}
+		// Keep the session visible while the blocking create request provisions it.
+		const placeholderId = `cloud-provisioning-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
+		const startedAt = new Date().toISOString();
+		this.pendingCreates.set(placeholderId, {
+			sessionId: placeholderId,
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts:1500]: Fixed in b091c72e4. The provisioning placeholder id is now generated with crypto.randomUUID() instead of Math.random().

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 4]
```diff
@@ -0,0 +1,2154 @@
...
+			metadata: {
+				origin: "cloud",
+				repoUrl: input.repoUrl,
+				git: { url: input.repoUrl, branch: input.branch ?? "" },
+				title: `Provisioning ${repositoryLabel(input.repoUrl)}…`,
+			},
+		});
+		// Show the placeholder before the next sidebar poll.
+		sendEvent(this.ctx, "chat_session_status", {
+			sessionId: placeholderId,
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts:1524]: Fixed in b091c72e4. The provisioning placeholder id is now generated with crypto.randomUUID() instead of Math.random().

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 5]
```diff
@@ -0,0 +1,2154 @@
...
+			// Swap an open placeholder thread to the real session.
+			sendEvent(this.ctx, "cloud_session_provisioned", {
+				placeholderId,
+				sessionId: String(created.sessionId ?? ""),
+			});
+			return created;
+		} finally {
+			this.pendingCreates.delete(placeholderId);
+			sendEvent(this.ctx, "chat_session_status", {
+				sessionId: placeholderId,
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts:1556]: Fixed in b091c72e4. The provisioning placeholder id is now generated with crypto.randomUUID() instead of Math.random().

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 6]
```diff
@@ -0,0 +1,2187 @@
...
+					this.ctx,
+					outerSessionId,
+					800,
+					messages,
+				),
+			});
+			const buffered = reconcileBufferedCloudEvents(
+				connection.bufferedEvents,
+				messages,
+				{ queueSnapshotApplied: Boolean(queueReply) },
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts]: Fixed in eb9590873. Rehydration now validates the pending-prompts reply before treating it as authoritative: only a successful reply that actually carries a prompts array is applied and allowed to supersede buffered queue events. Unsuccessful or malformed replies leave the queue untouched and replay the newest buffered queue snapshot. Added a manager-level regression test where the queue reply is malformed while a queue event sits in the reconnect buffer.

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 7]
```diff
@@ -0,0 +1,2195 @@
...
+							(!requestedBranch ||
+								session.repoContext.branch === requestedBranch) &&
+							Date.parse(session.createdAt) >= requestedAt - 5_000,
+					)
+					.sort(
+						(left, right) =>
+							Date.parse(right.createdAt) - Date.parse(left.createdAt),
+					)[0];
+				if (recovered) {
+					this.claimedSessionIds.add(recovered.id);
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts]: Addressed in 2e5a56509. For context: the previous code was already race-free in practice, because there is no await between the list result resolving and the claim add, so each recovery's check-and-claim ran atomically within one event-loop continuation. To make that invariant structural rather than incidental, candidate selection and claiming now happen inside a single synchronous helper (claimFirstUnclaimed) that can never be separated by an await, and the regression test now runs the two identical create requests truly concurrently with Promise.all and asserts they adopt distinct sessions.

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 8]
```diff
@@ -0,0 +1,2195 @@
...
+	}
+
+	/** Mirrors the authoritative queue reply into desktop state. */
+	private applyQueueSnapshot(
+		outerSessionId: string,
+		reply: { payload?: Record<string, unknown> },
+	): PromptInQueue[] {
+		const items = Array.isArray(reply.payload?.prompts)
+			? (reply.payload.prompts as Array<Record<string, unknown>>)
+			: [];
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts]: Fixed in 2e5a56509, following your suggestion: applyQueueSnapshot now throws on a reply without a prompts array, so the pending, update, and remove command paths surface an error instead of publishing an authoritative empty queue. Rehydration already routes invalid replies away from this helper and replays buffered queue state instead. Added a regression test proving a malformed reply rejects and leaves the existing queue intact.

File: apps/examples/desktop-app/sidecar/cloud-sessions.ts [hunk 9]
```diff
@@ -0,0 +1,2351 @@
...
+					signal: controller.signal,
+				},
+				input.organizationId?.trim()
+					? `${this.appBaseUrl}/dashboard/organization/integrations`
+					: undefined,
+				creationAuthToken,
+			);
+			if (created?.sessionId?.trim()) {
+				this.claimedSessionIds.add(created.sessionId.trim());
+			}
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/examples/desktop-app/sidecar/cloud-sessions.ts]: Fixed in f3db7a0a9. Valid catch, and the window is wider than a microtask: the server lists a session as soon as provisioning starts, minutes before the successful POST returns. create() now registers each in-flight request under a repo/model/branch/org key with a monotonic sequence, and timeout recovery waits for every earlier identical request to record its claim before adopting a listed candidate. Later peers wait on earlier ones only, so the waits cannot cycle, including when both requests fail into recovery. Added a regression test where a slow successful POST overlaps a fast-failing one: the recovery no longer steals the in-flight session and surfaces its own failure instead.

**Comment by saoudrizwan** [comment]
Did a full review of this branch: read the entire diff (SDK hub changes, sidecar, webview), ran the test suites, checked lint and merge cleanliness against latest main, and manually tested the toggle and onboarding flow end to end in a headless dev build. Overall verdict: this is in good shape architecturally, and I did not find any correctness bugs. Notes below.

## What I verified

- Desktop app: `bun run typecheck` passes. Full vitest suite is 614/615 with the only failures being `scripts/generate-update-manifest.test.ts` (imports `bun:test`, cannot run under vitest) and `webview/styles/theme-integration.test.ts` (Tailwind version drift). Neither file is touched by this PR, matching the description's claim that they are pre-existing.
- SDK: `@cline/core` unit tests pass except the known cloud-VM `workspace-manifest` git-remote artifact and one `checkpoint-restore` timeout that passes in isolation (flake under parallel load). The new `connection-headers.test.ts` and extended `boundary.test.ts` pass (39 tests).
- New desktop tests all pass: `cloud-sessions.test.ts` (70 behavior-level tests), `desktop-settings`, `feature-flags`, `commands-settings`, `use-chat-session`, `cloud-repositories`, `cloud-session-error` (125 tests total).
- Biome check on all changed areas: clean. The branch also merges cleanly with current `main` (12 commits ahead of the merge base, no conflicts).
- Manual E2E (headless `dev:sidecar` + `dev:web`, signed out): toggle appears in Settings > General with Preview badge, default off; enabling it makes the Local/Cloud picker appear in the composer live with no reload; clicking Cloud shows the signed-out onboarding panel with the sign-in CTA and three steps; disabling the toggle snaps the composer back to Local. The setting persists to `~/.cline/data/settings/code-settings.json` and the sidecar log stays clean (no error spam from the signed-out discovery path).

## Architecture assessment

- The layering is right. The generic capability (header-authenticated hub connections, `approval.list_pending`) went into `@cline/core`/`@cline/shared`, while everything cloud-session-specific stayed in the desktop app's sidecar. `approval.list_pending` is capability-gated and the handler enforces the same participant check as other session commands, including a stranger-client rejection test.
- The `NodeHubClient` changes are the riskiest part of the PR since every hub consumer shares that code, and they hold up well: `authToken` and `resolveConnectionHeaders` are mutually exclusive (checked in both constructor and connect), header resolution is timeout-bounded, a generation counter plus socket-identity guards prevent a stale connect attempt from clobbering a newer one after `close()`, and a socket that opened but failed registration is closed rather than left half-alive. The registry-token skip for header-auth clients closes a subtle loopback-token leak. The 440-line test file covers these races directly.
- Routing by `ses-` prefix for cloud outer session ids is safe: local ids are `${Date.now()}_<nanoid>` from `createSessionId`, and the webview's planned ids are `session_...`, so no collision is possible.
- The settings placement is well reasoned: a desktop-owned `code-settings.json` avoids the strict-schema stripping problem an older CLI writer would cause in `global-settings.json`, writes are atomic (temp file + rename), and the gate deliberately only blocks new session creation so a rollback never strands an existing cloud session.
- The failure-mode thinking throughout `cloud-sessions.ts` is unusually thorough for a first shippable cut: create-timeout recovery with claim ordering so two identical concurrent creates cannot adopt the same sandbox, event buffering plus per-run-segment reconciliation during rehydration, expiry reaping so a dead sandbox does not reconnect-loop, approval handling that leaves cloud approvals pending on local shutdown (the pod keeps running), and tombstoning during delete. Each nontrivial decision has a comment explaining why, and almost every one has a matching test.

## Non-blocking notes

1. API load: `list_chat_sessions` merges cloud discovery on every sidebar poll (12s) for every signed-in user, regardless of the toggle. That is a deliberate rollback-safety choice (existing sessions stay listed when the flag is off) and it is bounded by the 2s budget with a cached-list fallback, but it does mean roughly five extra `/api/v1/session` calls per minute per signed-in desktop user plus a 60s-cached org lookup, even for users who never opted in. If backend load matters at rollout, a cheap improvement would be to skip the remote fetch when the toggle is off and the manager has never seen a cloud session for this account.
2. `settings-view.tsx` carries a commented-out PostHog gate block for the future flag that controls the toggle's visibility. The comment explains the plan well, but it is still dead code in the tree; a follow-up issue reference would let it be dropped.
3. `cloud-sessions.ts` is a single 2,400-line module holding the REST client, the connection manager, and the event-reconciliation logic, mirrored by a 2,500-line test file. It is cohesive and readable today, but if it grows again it would be worth splitting into api/manager/reconcile modules. Same tendency in `use-chat-session.ts`, which is now ~2,700 lines, though the new cloud merge logic is at least extracted into a pure, tested function.
4. Tiny duplication: the "owner/repo" label parsing appears in three places (sidecar `repositoryLabel`, `chat-input-bar` context label, `page.tsx` provisioning phase). Could be one helper in `cloud-repositories.ts`.
5. In `CloudSessionApi.create()` timeout recovery, the returned `cleanupAuthToken` is the original creation token even though recovery listed with a fresh one; after a ~10 minute create the cleanup token may be expired. Only affects the rare dispose-during-provisioning cleanup path, which already logs and swallows failures.

None of these should block merging. Nice work on this one, especially the test discipline on the concurrency paths.

**Comment by saoudrizwan** [comment]
Follow-up: I ran a second, much deeper review pass with parallel focused audits (hub client concurrency, cloud session manager, webview state machine, security, onboarding/picker flows, and test quality including mutation testing). Several findings below were proven with runtime reproductions before reporting; those are marked "reproduced". Reproduction test files were throwaway and are not committed. Ordered by priority.

## Should fix before merge

**1. `NodeHubClient` zombie-connection race (reproduced).** `sdk/packages/core/src/hub/client/index.ts`. Every mutation window in the new connect machinery is generation-guarded except the post-registration continuation. If `close()` runs in the microtask window after the `client.register` reply resolves but before the continuation runs, the continuation sets `registered = true` on a closed client. That stale flag then poisons the catch guard in `connect()` (`if (!this.registered && ...)`): on a later reconnect whose registration is rejected (for example token expiry), the unregistered socket is not closed, `isConnected()` returns true, and every subsequent `connect()` early-returns on the open socket without ever re-registering. Commands flow from a permanently unregistered client. A deterministic repro asserted all four consequences. Fix direction: capture the generation in `connect()` and abort the continuation when `generation !== this.connectGeneration || this.closedByClient`, and drop the `!this.registered` condition from the catch guard (the socket-identity check alone is sufficient and cannot be poisoned). This client is shared by the CLI, VS Code extension, and desktop, so it is the highest-blast-radius item here.

**2. `CloudBranchPicker` infinite scroll permanently deadlocks (reproduced).** `webview/components/views/chat/welcome-workspace-controls.tsx`. `loadMore`'s `finally` only resets `loadingMore` when `requestKeyRef.current` still matches the key captured at call time. If the user types a search character while a page-2 fetch is in flight, the key changes, the stale fetch settles without resetting the flag, and nothing else ever does. From then on the sentinel shows "Loading more branches..." forever and both the observer effect and `loadMore` short-circuit on `loadingMore`. On repos with more than one page of branches, later pages become unselectable for the life of the welcome screen. Fix: reset `loadingMore` unconditionally in the `finally`.

**3. First cloud prompt's optimistic bubble is invisible to the snapshot merge (reproduced).** `webview/hooks/use-chat-session.ts`. `mergeCloudSnapshotWithLive` drops any message whose `sessionId` differs from the target session before consulting `optimisticStates`. The first prompt of a fresh cloud session keeps the client-planned `plannedSessionId` on the message, while the sidecar returns a server-assigned id; the post-start fixup re-points the optimistic-state map entry but not the message itself. Result: the designed retention semantics (keep pending bubbles the snapshot has not reflected, always keep failed bubbles) never apply to the first prompt. Reproduced two user-visible cases: a lagging snapshot merges to a transcript with zero user messages, and a failed first send loses its prompt bubble on the next rehydration, leaving an error banner with no prompt above it. The happy path usually masks this because the sidecar appends the confirmed prompt to `live.messages`. Fix direction: re-key the optimistic message's `sessionId` (a `setMessages` + update next to the existing `cloudOptimistic.sessionId` fixup).

**4. Create-recovery can double-adopt in the uncovered direction.** `sidecar/cloud-sessions.ts`. The claim ordering makes a recovering create wait only for peers with a smaller sequence. The inverted interleaving is unguarded: request A (seq 1) fails fast with `request_failed` (any non-401/404/410/412 status qualifies, including an instant 500, not just the 610s timeout), while identical request B (seq 2) is still POSTing but its record is already listed. A waits on nothing, lists, and claims B's session; B's success path then adds the same id to `claimedSessionIds` without checking whether a recovery already took it. Both callers resolve with the same sandbox (reproduced: both returned the same session id). Two adjacent gaps with the same mechanism: a branchless create and a branch-specific create hash to different claim keys but the branchless recovery filter skips the branch match, so they can cross-adopt with no ordering protection at all; and claims are per-process while the list is account-wide with 60 seconds of backward slack, so a fast-failing create can adopt an identical-config session started by another device (and the first-prompt auto-titling would then rename it). Suggested hardening: have the success path treat an already-claimed id as a conflict, make recovery also wait on later in-flight peers for the same key, include branchless requests in branch-specific claim keys, and consider tightening the time window.

**5. Org switch can wipe a repository the user just picked.** `webview/components/views/chat/welcome-chat.tsx`. The stale-repo-clearing effect compares `repoUrl` against `cloudSetup.repositoryUrls`, but that snapshot only refreshes on mount, account-id change, window focus, or the 6s poll (which stops in "ready" status). Switching the active organization in Settings changes none of those, and `welcome-chat` does not subscribe to the `cloud_sessions_changed` broadcast (only `use-session-history` does). So after an in-app org switch the user can pick a new-org repository from the correctly scoped picker and the effect immediately clears it against the old org's list. `CloudRepositoryPicker.onRepositoriesLoaded` also never feeds its fresh list back into `cloudSetup.repositoryUrls`, which is what makes this deterministic rather than a race. Fix: refresh `cloudSetup` from picker loads and/or subscribe to `cloud_sessions_changed`.

## Security

**6. Spoofable cloud-error envelope can aim the "Connect GitHub" button at an attacker URL (medium).** The `CLOUD_SESSION_ERROR:{json}` envelope travels in `Error.message`, and the webview authenticates it by string prefix only. Hub command failures against the pod carry the pod's reply string verbatim (`resolved.error.message` in the hub client) through the sidecar transport into the webview error state, where `parseCloudSessionError` will happily parse a pod-crafted `{"code":"github_not_connected","connectUrl":"https://attacker.example/..."}` and render a trusted-looking "Connect GitHub" button; `open_external_url` validates protocol but not origin. Cloud sandboxes execute untrusted repo code by design, so this is a credible phishing vector requiring one user click. Fix: validate `connectUrl` against the known `appBaseUrl` origin before rendering the action button (and ideally only honor `connectUrl` on sidecar-constructed errors). The rest of the security surface checked out clean: every path interpolation is `encodeURIComponent`-wrapped, tokens never reach logs/telemetry/URLs, header auth is mutually exclusive with subprotocol auth and rejects `Sec-WebSocket-Protocol`, remote strings render as React text with no `dangerouslySetInnerHTML`, and `open_external_url` uses a strict protocol allowlist with the Windows `rundll32` opener.

## Worth fixing, lower priority

- `cloud_session_sync_failed` handler in `use-chat-session.ts` calls `setError(record.message)` without `humanizeCloudSessionError`, so a signed-out user with an open cloud thread sees the raw JSON envelope in the error banner. Also every expected-state `CloudSessionError` (expired attach, signed-out send) lands in error telemetry with the machine envelope as the message.
- Delete is offered on provisioning placeholders in both the sidebar and sessions-view menus (the confirm dialog even promises deletion) but the sidecar always rejects it; the toast is humanized, but the affordance should be disabled like Rename already is.
- With the `CLINE_CODE_CLOUD_AGENTS` env override set, the Settings toggle silently diverges from effective behavior in both directions (toggle on, feature off; toggle off, feature on) with no hint. Consider surfacing the effective value from `get_feature_flags`.
- The slash-command menu is not cloud-gated even though the cloud placeholder text says slash commands are local-only; typing `/` in a cloud session offers local skills/workflows the sandbox cannot resolve (mentions are correctly gated).
- `ModelSelector`'s auto-correct effect ignores `cloudSettingsLocked`; if an attached cloud session's model id is missing from the local catalog, it silently switches the model and the next send applies it remotely via `session.update_connection`, contradicting the "model is fixed" tooltip.
- A session deleted from another device reconnect-loops every ~5s indefinitely (upgrade attempt, header resolve, then a REST list from the failure hook) until app restart. `disposeConnectionIfSessionGone` already has the information to distinguish "list succeeded and the session is absent" from "offline"; only the expired case currently reaps.
- `resetCloudSessionManager` nulls `ctx.cloudSessionManager` before awaiting `dispose()`, so a command arriving in the dispose window creates a fresh manager whose just-created `liveSessions`/`pendingApprovals` entries the old dispose then deletes. Narrow window, but `save_provider_settings` for cline triggers it on every save.
- The `recoveredAfterDisconnect` branch maps unknown statuses to "completed" while the `cloud_session_rehydrated` handler deliberately leaves unknown statuses alone; pick the conservative behavior for both. Also `chat_session_ended` uses reason "failed" from rehydration but "error" from the live path, so the same terminal state lands as two different statuses.
- Minor hub-client notes: a stale attempt's late timeout/error handlers overwrite `lastCloseError` before the identity guard (misreporting only), and `close()` now always resets `lastCloseError` to the generic message even with no socket, wiping the real failure cause for `getConnectionError()` readers.

## Test quality (mutation audit)

Overall strong: 5 of 8 targeted mutations were caught by exactly the test whose name claims the behavior (claim ordering, claim set, newest-buffered-queue replay, the connect generation guard, expiry semantics). The concentrated gap is `mergeCloudSnapshotWithLive`: three mutations survived (the `budget > 0` guard, the entire count-delta arithmetic, and the error-bubble exemption in the drop path). The test named "reconciles identical optimistic prompts by authoritative count delta" passes with the delta arithmetic deleted, so the budget mechanism is effectively unpinned; fixing finding 3 should come with tests that pin it. Other coverage gaps worth closing: the webview rehydrated status-mapping branches, the `cloud_session_sync_failed` handler, busy-state ownership for queue/steer sends, and the buffer/dedupe caps. One CI-flake candidate: the reconnect header-refresh test races a real jittered reconnect timer against a 3s deadline over a real socket.

## Re-verified safe

For completeness, the second pass specifically re-verified and cleared: concurrent `connect()` dedupe and all close-during-resolver interleavings other than finding 1; `ensureConnection` dedupe and failure cleanup; delete tombstoning vs concurrent attach; rehydration single-flight and rerun bounding; event-buffer flip ordering; StrictMode double-invocation safety of the optimistic bookkeeping; cross-session bleed on fast thread switches; the local-session status-dedupe change in `handleHubLiveEvent` (no local regressions found); `approval.list_pending` authz (stranger client rejected, payload consistent with the event); and the settings toggle save/rollback/broadcast path. One earlier hypothesis was checked and found invalid: hub status vocabulary cannot deliver "cancelled" to the webview mappers (the hub maps local cancelled to aborted), so both mappers do cover the real vocabulary.

Happy to turn any of items 1-6 into fix PRs.

**Comment by saoudrizwan** [comment]
Pushed fixes for everything from the deep-review pass, `e03b4577b` through `da759853e`. Each fix ships with a regression test that was verified to fail against the unfixed code before committing.

**Must-fix findings**

- `fix(sdk)`: closed the hub client zombie-connection race. The post-registration continuation is now generation-guarded, so a `close()` racing the register reply aborts the attempt instead of marking a closed client registered; the `connect()` catch guard now keys on socket identity alone. Also hardened: stale attempts' late timeout/error/close handlers no longer clobber `lastCloseError` for a newer attempt, and `close()` with no socket preserves the real failure cause. New test drives the exact reply-raced-close interleaving plus the follow-on rejected-registration and fresh-reconnect stages.
- `fix(desktop)`: branch picker pagination no longer deadlocks when the search query changes while a page fetch is in flight; `loadingMore` resets unconditionally (only one page fetch can ever be in flight).
- `fix(desktop)`: the first cloud prompt's optimistic bubble is re-keyed to the server-assigned session id after create, so the snapshot merge's cross-session guard can no longer drop it. Also added the previously missing tests pinning the merge's count-delta budget (zero-budget retention, one-consumption-per-new-copy) and error-bubble preservation, which mutation testing showed were unpinned.
- `fix(desktop)`: create-recovery claim race closed in both directions. In-flight peers are now keyed by repo/model/org (branch excluded so branchless recoveries see branch-specific peers), peer entries settle when their POST settles (never after recovery, so waits cannot cycle), recovery waits on all other peers re-snapshotting until stable, and recovery only triggers on timeout/5xx/no-status failures; a fast 4xx no longer risks adopting an identical-config session created by another device. Three new tests cover the inverted direction, the cross-key case, and the 4xx no-recovery gate.
- `fix(desktop)`: the cloud setup snapshot refreshes from the repository picker's own loads and from the `cloud_sessions_changed` broadcast, so an in-app org switch can no longer wipe a repository just picked from the new scope.
- `fix(desktop)` (security): `parseCloudSessionError` now drops `connectUrl` values whose origin is not a known Cline app base URL, so a pod-spoofed error envelope cannot aim the Connect GitHub button at an attacker page.

**Lower-priority cleanups (two commits)**

Sidecar: connections whose session vanished from a successful list are reaped instead of redialing a dead proxy every ~5s forever; `dispose()` clears desktop-visible state synchronously before its awaits so a manager rebuilt mid-dispose is not clobbered; rehydrated failed runs report the same `chat_session_ended` reason as the live path.

Webview: the sync-failed banner and rehydration fallback humanize cloud error envelopes (and the fallback is scoped to the active session); the recovered-send status mapper covers `cancelled` and leaves unknown statuses alone instead of defaulting to completed; Delete is disabled on provisioning placeholders in both menus; Settings shows a note when the `CLINE_CODE_CLOUD_AGENTS` override makes the toggle diverge from effective behavior (and the settings sections load concurrently); the slash-command menu is gated to local sessions like mentions; the model selector no longer silently "corrects" a locked cloud session's model when its id is missing from the local catalog (one existing test asserted the buggy behavior and was updated).

**Verification**

- Desktop: `bun run typecheck` passes; full vitest is 632/633 with only the two known pre-existing failures (`scripts/generate-update-manifest.test.ts`, `webview/styles/theme-integration.test.ts`). All 170 tests across the touched cloud files pass.
- SDK: all 208 hub tests pass; full `@cline/core` unit run shows only the two known environment artifacts (git `insteadOf` rewrite, and a checkpoint-restore timeout flake that passes standalone).
- Biome is clean on all touched files. The one repo-level format diagnostic (`sdk/packages/core/src/hub/daemon/entry.ts`) comes from the latest main merge, untouched here.
- Manual smoke test against the running dev app (headless sidecar + web UI, signed out): toggle on/off, Local/Cloud picker appearing live, signed-out onboarding panel, and snap-back to local all still work; no console errors.
